src/frequenz/sdk/data_ingestion/resampling/timeseries_resampler.py

```diff
@@ -0,0 +1,118 @@
+        """
+        self._period_s = period_s
+        self._max_data_age_in_periods: float = max_data_age_in_periods
+        self._resampling_function: ResamplingFunction = resampling_function
+        self._resamplers: Dict[TimeSeriesId, IndividualTimeSeriesResampler] = {}
+
+    def add_time_series(self, time_series_id: TimeSeriesId) -> None:
+        """Create a new resampler for a specific time series.
+
+        Args:
+            time_series_id: time series id
+
+        Raises:
+            ValueError: if resampler for the provided timer_series_id already exists
+        """
+        resampler = self._resamplers.get(time_series_id)
+        if resampler is None:
+            raise ValueError(
+                f"Resampler for time series {time_series_id} already exists!"
+            )
+
+        self._resamplers[time_series_id] = IndividualTimeSeriesResampler(
+            buffer_size=DEFAULT_BUFFER_SIZE,
+            period_s=self._period_s,
+            max_data_age_in_periods=self._max_data_age_in_periods,
+            resampling_function=self._resampling_function,
+        )
+
+    def remove_timeseries(self, time_series_id: TimeSeriesId) -> None:
+        """Remove a resampler for a specific time series.
+
+        Args:
+            time_series_id: time series id, for which to remove the resampler
+
+        Raises:
+            ValueError: if resampler for the provided timer_series_id doesn't exist
+        """
+        resampler = self._resamplers.get(time_series_id)
+        if resampler is None:
+            raise ValueError(f"No resampler for time series {time_series_id} found!")
+
+        del self._resamplers[time_series_id]
+
+    def add_sample(self, time_series_id: TimeSeriesId, sample: Sample) -> None:
+        """Add a sample for a specific time series.
+
+        Args:
+            time_series_id: time series id, for which to remove the resampler
+            sample: sample to be added
+
+        Raises:
+            ValueError: if resampler for the provided timer_series_id doesn't exist
+        """
+        resampler = self._resamplers.get(time_series_id)
+        if resampler is None:
+            raise ValueError(f"No resampler for time series {time_series_id} found!")
+
+        resampler.add_sample(sample)
+
+    def resample(self) -> Dict[TimeSeriesId, Optional[Sample]]:
+        """Resample samples for all time series.
+
+        Returns:
+            mapping of time series ids and their newly resampled samples
+        """
+        now = datetime.now(tz=pytz.UTC)
+        samples: Dict[TimeSeriesId, Optional[Sample]] = {}
+        for time_series_id, resampler in self._resamplers.items():
+            value: Optional[float] = resampler.resample()
+            if value is not None:
+                samples[time_series_id] = Sample(timestamp=now, value=value)
+            else:
+                samples[time_series_id] = None
+        return samples
```

Comment:
Looking at this, I think it should be a generator, otherwise we are:

1. Copying a lot of data
2. Delaying the sending of samples until all samples were calculated

```suggestion
    def resample(self) -> Iterator[Tuple[TimeSeriesId, Sample]]:
        now = datetime.now(tz=pytz.UTC)
        return map(lambda i: (i[0], Sample(now, i[1].resample())),
            self._resamplers.items())
```

As a perk the implementation becomes trivial, although a bit cryptic, it could also be something like this instead of the map:

```py
        for time_series_id, resampler in self._resamplers.items():
            yield timeseries_id, Sample(timestamp=now, value=resampler.resample())
```

Reply:
That's a neat piece of code, but I cannot get the types to work. I tried the following return types:
- `Iterator[Tuple[TimeSeriesId, Sample]]`
- `Generator[Tuple[TimeSeriesId, Sample]], None, None]
- `Iterable[Tuple[TimeSeriesId, Sample]]`

and I am always getting:
```
(mos-new) ➜  frequenz-sdk-python git:(resampling-actor-revised) ✗ darglint src examples
src/frequenz/sdk/data_ingestion/resampling/timeseries_resampler.py:resample:111: DAR301: - yield
src/frequenz/sdk/data_ingestion/resampling/timeseries_resampler.py:resample:114: DAR202: + return
```

Reply:
Which implementation did you go for?

Reply:
@leandro-lucarella-frequenz the one you suggested. I was just missing the `Yields:` instead of `Returns:` in the docstring. :)

Reply:
I suggested 2 :smile:, but never mind if it worked :)